### PR TITLE
Proxy Protocol: Pass through V2 TLVs to upstream

### PR DESCRIPTION
**Pull request by botengyao**
Signed-off-by: Boteng Yao <boteng@google.com>
Co-authored-by: Siyuan Zhou <siyuan.zhou@mongodb.com>

This PR will store Proxy Protocol V2 TLVs to filter state based on the API config. And upstream can send these TLVs in transport socket based on the pass through config as well. 

Some discussion already happened here: #18940 
And thanks @siyuan.zhou@mongodb.com for the initial work.

Commit Message:
Additional Description:
Risk Level: Low
Testing: unit & integration test
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by botengyao** [comment]
Pending release note.

**Comment by botengyao** [comment]
/assign @yanavlasov 
Thanks!

**Review by yanavlasov — COMMENTED** [review]
/wait

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.h [hunk 1]
```diff
@@ -69,6 +74,8 @@ class Config : public Logger::Loggable<Logger::Id::filter> {
 private:
   absl::flat_hash_map<uint8_t, KeyValuePair> tlv_types_;
   const bool allow_requests_without_proxy_protocol_;
+  bool pass_all_tlvs_ = false;
```
Comment (yanavlasov) [inline source/extensions/filters/listener/proxy_protocol/proxy_protocol.h:77]: Please make config members `const`.

**Review by yanavlasov — CHANGES_REQUESTED** [review]

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.h [hunk 2]
```diff
@@ -37,6 +37,8 @@ class UpstreamProxyProtocolSocket : public TransportSockets::PassthroughSocket,
   Network::TransportSocketCallbacks* callbacks_{};
   Buffer::OwnedImpl header_buffer_{};
   ProxyProtocolConfig_Version version_{ProxyProtocolConfig_Version::ProxyProtocolConfig_Version_V1};
+  bool pass_all_tlvs_ = false;
```
Comment (yanavlasov) [inline source/extensions/transport_sockets/proxy_protocol/proxy_protocol.h:40]: If these values are immutable, please make them `const`.

**Review by yanavlasov — CHANGES_REQUESTED** [review]

File: test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc [hunk 3]
```diff
@@ -398,5 +400,239 @@ TEST_P(ProxyProtocolHttpIntegrationTest, TestProxyProtocolHealthCheck) {
...
+                         TestUtility::ipTestParamsToString);
+
+// This test adding the listener proxy protocol filter and upstream proxy filter, the TLVs
+// are passed by listener and re-generated in transport socket based on API config.
+TEST_P(ProxyProtocolTLVsIntegrationTest, TestV2TLVProxyProtocol_PassSepcificTLVs) {
+  setup(false, {0x05, 0x06}, {0x06});
+  initialize();
+
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort("listener_0"));
+  ;
```
Comment (yanavlasov) [inline test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc:490]: Empty expression.

File: test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc [hunk 4]
```diff
@@ -398,5 +400,239 @@ TEST_P(ProxyProtocolHttpIntegrationTest, TestProxyProtocolHealthCheck) {
...
+  std::vector<uint8_t> tlvs_upstream_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, ProxyProtocolTLVsIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+// This test adding the listener proxy protocol filter and upstream proxy filter, the TLVs
+// are passed by listener and re-generated in transport socket based on API config.
+TEST_P(ProxyProtocolTLVsIntegrationTest, TestV2TLVProxyProtocol_PassSepcificTLVs) {
```
Comment (yanavlasov) [inline test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc:485]: You do not need to add `_` in tests names per style.

**Review by yanavlasov — CHANGES_REQUESTED** [review]

File: test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc (hunk 4 above)
Comment (botengyao) [reply test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc:485]: Done.

File: test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc (hunk 3 above)
Comment (botengyao) [reply test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc:490]: Thanks, done.

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.h (hunk 2 above)
Comment (botengyao) [reply source/extensions/transport_sockets/proxy_protocol/proxy_protocol.h:40]: It is mutable at beginning based on the config match type. It is related to API setting.. cc @envoyproxy/api-shepherds as well.

File: envoy/network/proxy_protocol.h [hunk 5]
```diff
@@ -1,13 +1,25 @@
...
+#include <vector>
+
 #include "envoy/network/address.h"
 
 namespace Envoy {
 namespace Network {
 
+struct ProxyProtocolTLV {
+  const uint8_t type;
+  const std::string value;
```
[the remark this answers is not in the document]
Comment (botengyao) [reply envoy/network/proxy_protocol.h:14]: Done.

File: source/extensions/common/proxy_protocol/proxy_protocol_header.cc [hunk 6]
```diff
@@ -93,10 +98,48 @@ void generateV2Header(const std::string& src_addr, const std::string& dst_addr,
...
+      continue;
+    }
+    extension_length += PROXY_PROTO_V2_TLV_TYPE_LENGTH_LEN + tlv.value.size();
+    if (extension_length > std::numeric_limits<uint16_t>::max()) {
+      ExceptionUtil::throwEnvoyException(
+          fmt::format("proxy protocol TLVs exceed length limit {}, already got {}",
+                      std::numeric_limits<uint16_t>::max(), extension_length));
+    }
+  }
+  assert(extension_length <= std::numeric_limits<uint16_t>::max());
```
[the remark this answers is not in the document]
Comment (botengyao) [reply source/extensions/common/proxy_protocol/proxy_protocol_header.cc:127]: Thanks!

File: source/extensions/common/proxy_protocol/proxy_protocol_header.cc [hunk 7]
```diff
@@ -93,10 +98,48 @@ void generateV2Header(const std::string& src_addr, const std::string& dst_addr,
...
+void generateV2Header(const Network::ProxyProtocolData& proxy_proto_data, Buffer::Instance& out,
+                      bool pass_all_tlvs, const absl::flat_hash_set<uint8_t>& pass_through_tlvs) {
+  uint64_t extension_length = 0;
+  for (auto&& tlv : proxy_proto_data.tlv_vector_) {
+    if (!pass_all_tlvs && !pass_through_tlvs.contains(tlv.type)) {
+      continue;
+    }
+    extension_length += PROXY_PROTO_V2_TLV_TYPE_LENGTH_LEN + tlv.value.size();
+    if (extension_length > std::numeric_limits<uint16_t>::max()) {
+      ExceptionUtil::throwEnvoyException(
```
[the remark this answers is not in the document]
Comment (botengyao) [reply source/extensions/common/proxy_protocol/proxy_protocol_header.cc:122]: Done.

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.h (hunk 2 above)
Comment (botengyao) [reply source/extensions/transport_sockets/proxy_protocol/proxy_protocol.h:53]: We can move to the method defination like, 
```
pass_all_tlvs_(proto_config.has_pass_through_tlvs() ? proto_config.pass_through_tlvs().match_type() == ProxyProtocolPassThroughTLVs::INCLUDE_ALL
                     : false)
```
But the we still need to check the `INCLUDE` type, and looks like the current is cleaner, WDYT?

**Comment by botengyao** [comment]
Hi @envoyproxy/api-shepherds, and @markdroth, can we get this API reviewed? Thanks!

**Comment by yanavlasov** [comment]
@botengyao looks like build failure is real: https://source.cloud.google.com/results/invocations/f7ace5fd-8582-499c-aa4d-36e93034a549/log

**Comment by botengyao** [comment]
Hi @yanavlasov, @markdroth, and @envoyproxy/api-shepherds PTAL, thanks!

File: test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc [hunk 8]
```diff
@@ -398,5 +400,238 @@ TEST_P(ProxyProtocolHttpIntegrationTest, TestProxyProtocolHealthCheck) {
...
+  tcp_client->close();
+  ASSERT_TRUE(fake_upstream_connection_->waitForDisconnect());
+}
+
+TEST_P(ProxyProtocolTLVsIntegrationTest, TestV2TLVProxyProtocolPassAll) {
+  setup(true, {}, {});
+  initialize();
+
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort("listener_0"));
+  ;
```
Comment (yanavlasov) [inline test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc:560]: nit: extra `;`

**Review by yanavlasov — APPROVED** [review]
LGTM

**Review by suniltheta — COMMENTED** [review]
Just 1 clarifying question

File: api/envoy/config/core/v3/proxy_protocol.proto [hunk 9]
```diff
@@ -23,4 +43,8 @@ message ProxyProtocolConfig {
 
   // The PROXY protocol version to use. See https://www.haproxy.org/download/2.1/doc/proxy-protocol.txt for details
   Version version = 1;
+
+  // This config controls which TLVs can be passed to filter state if it is Proxy Protocol
+  // V2 header. If there is no setting for this field, no TLVs will be passed through.
```
[the remark this answers is not in the document]
Comment (botengyao) [reply api/envoy/config/core/v3/proxy_protocol.proto:48]: For proto3, there is a difference between not setting and default, right? All configs are no setting before this PR, and they will keep the original behavior without any change - not passing TLVs.

[the remark this answers is not in the document]
Comment (suniltheta) [reply api/envoy/config/core/v3/proxy_protocol.proto:48]: understood. thanks for clarifying about backward compatibility.

[the remark this answers is not in the document]
Comment (suniltheta) [reply api/envoy/config/core/v3/proxy_protocol.proto:48]: 1 final comment, we can update the change logs about the new config. That is missing on this PR.

**Pull request merged**